vertical-pod-autoscaler/enhancements/4016-in-place-updates-support/README.md

```diff
@@ -17,88 +17,135 @@
+Today, only the VPA admission controller is responsible for changing the pod spec.
+
+The VPA updater is responsible for evicting pods so that the admission controller can change them
+during admission.
+
+In the newly added `InPlaceOrRecreate` mode, the VPA Updater will attempt to execute in-place
+updates FIRST. In certain situations, if it is unable to process an in-place update in time, it
+will evict the pod to force a change.
 
-In the newly added `InPlaceOnly` and `InPlaceOrRecreate` modes VPA Updater will execute in-place updates (which change
-pod spec without involving admission controller).
+We classify three types of updates in the context of this new mode:
 
-### Applying Updates During Pod Admission
+1. Updates on pod admission
+2. Disruption-free updates
+3. Disruptive updates
 
-For VPAs in `InPlaceOnly` and `InPlaceOrRecreate` modes VPA Admission Controller will apply updates to starting pods,
-like it does for VPAs in `Initial`, `Auto`, and `Recreate` modes.
+**NOTE:** Here "disruptive" is from the perspective of VPA. As mentioned above, the container
+runtime MAY choose to restart a container or pod as needed.
 
-### Applying Disruption-free Updates
+More on these two types of changes below.
 
-When an update only changes resources for which a container indicates that it doesn't require a restart, then VPA can
-attempt to actuate the change without disrupting the pod. VPA Updater will:
+### 1. Applying Updates During Pod Admission
+
+For VPAs using the new `InPlaceOrRecreate` mode, the VPA Admission Controller will apply updates to
+starting pods just as it does for VPAs in `Initial`, `Auto`, and `Recreate` modes.
+
+### 2. Applying Disruption-free Updates (**NEW**)
+
+In the `InPlaceOrRecreate` mode when an update only changes resources for which a container
+indicates that it doesn't require a restart, then VPA can attempt to actuate the change without
+disrupting the pod. VPA Updater will:
 * attempt to actuate such updates in place,
 * attempt to actuate them if difference between recommendation and request is at least 10%
   * even if pod has been running for less than 12h,
 * if necessary execute only partial updates (if some changes would require a restart).
 
-### Applying Disruptive Updates
+If VPA updater fails to update the size of a container, we ignore the failure and try again. We
+will not intentionally evict the pod unless the conditions call for a disruptive update (see
+below).
 
-In both `InPlaceOnly` and `InPlaceOrRecreate` modes VPA updater will attempt to apply updates that require container
-restart in place. It will update them under conditions that would trigger update with `Recreate` mode. That is it will
-apply disruptive updates if:
+### 3. Applying Disruptive Updates
+
+In the `InPlaceOrRecreate` modes VPA updater will attempt to apply updates that require container
+restart in place. It will update them under the same conditions that would trigger update with
```

Comment:
```suggestion
In the `InPlaceOrRecreate` modes, and for updates that require a container restart, the VPA updater will attempt to apply updates in place. It will update them under the same conditions that would trigger an update with
```

Reply:
Thank you. Done.